packages/apidom-parser/src/parser.ts

```diff
@@ -18,119 +17,124 @@ export type {
-      const matches = source.match(detectionRegExp);
+    const { detectionRegExp } = adapter;
+    const matches = source.match(detectionRegExp);
 
-      if (matches === null) {
-        return new MediaTypes().unknownMediaType;
-      }
+    if (matches === null) {
+      return new MediaTypes().unknownMediaType;
+    }
 
-      const { groups } = matches;
-      const version = groups?.version || groups?.version_json || groups?.version_yaml;
-      const format = groups?.version_json ? 'json' : groups?.version_yaml ? 'yaml' : 'generic';
+    const { groups } = matches;
+    const version = groups?.version || groups?.version_json || groups?.version_yaml;
+    const format = groups?.version_json ? 'json' : groups?.version_yaml ? 'yaml' : 'generic';
 
-      if (typeof version === 'undefined') {
-        return adapter.mediaTypes.latest();
-      }
+    if (typeof version === 'undefined') {
+      return adapter.mediaTypes.latest();
+    }
 
-      // @ts-ignore
-      return adapter.mediaTypes.findBy(version, format);
-    };
-
-    this.parse = async function parse(source: string, options: ApiDOMParserOptions = {}) {
-      let adapter;
-
-      try {
-        adapter = await findAdapter(source, options.mediaType);
-      } catch (error: unknown) {
-        throw new ParserError(
-          'Encountered an unexpected error while matching parser adapters against the source.',
-          {
-            source,
-            parserOptions: options,
-            cause: error,
-          },
-        );
-      }
+    // @ts-ignore
+    return adapter.mediaTypes.findBy(version, format);
+  }
 
-      if (isUndefined(adapter)) {
-        throw new ParserError('Source did not match any registered parsers', {
-          source,
-          parserOptions: options,
-        });
-      }
+  async parse(source: string, options: ApiDOMParserOptions = {}): Promise<ParseResultElement> {
+    let adapter;
 
-      try {
-        return adapter.parse(source, options);
-      } catch (error: unknown) {
-        throw new ParserError('Parsing encountered an unexpected error.', {
+    try {
+      adapter = await this.findAdapter(source, options.mediaType);
+    } catch (error: unknown) {
+      throw new ParserError(
+        'Encountered an unexpected error while matching parser adapters against the source.',
+        {
           source,
           parserOptions: options,
           cause: error,
-        });
-      }
-    };
-  },
-);
+        },
+      );
+    }
+
+    if (isUndefined(adapter)) {
+      throw new ParserError('Source did not match any registered parsers', {
+        source,
+        parserOptions: options,
+      });
+    }
+
+    try {
+      return adapter.parse(source, options);
+    } catch (error: unknown) {
+      throw new ParserError('Parsing encountered an unexpected error.', {
+        source,
+        parserOptions: options,
+        cause: error,
+      });
+    }
+  }
+}
+
+function ApiDOMParser(): ApiDOMParserClass {
```

Comment:
We can directly define a constructor within the `ApiDOMParser` class and use [target.new](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/new.target) meta-property to check for presence of `new` keyword.